README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# ScatMatch <img src="man/figures/scatmatch_logo.png" align="right" style="padding-left:10px;background-color:white;" />

<!-- badges: start -->

[![Project Status: Active â€“ The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://www.tidyverse.org/lifecycle/#stable)
[![GitHub
issues](https://img.shields.io/github/issues/bartesto/ScatMatch.svg?style=popout)](https://github.com/bartesto/ScatMatch/issues/)
[![Last-changedate](https://img.shields.io/github/last-commit/bartesto/ScatMatch.svg)](https://github.com/bartesto/ScatMatch/commits/master)
[![Travis build
status](https://travis-ci.org/bartesto/ScatMatch.svg?branch=master)](https://travis-ci.org/bartesto/ScatMatch)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/bartesto/ScatMatch?branch=master&svg=true)](https://ci.appveyor.com/project/bartesto/ScatMatch)
[![Github top
language](https://img.shields.io/github/languages/top/bartesto/ScatMatch.svg)](https://github.com/bartesto/ScatMatch/)
[![DOI](https://zenodo.org/badge/202643428.svg)](https://zenodo.org/badge/latestdoi/202643428)
[![R-CMD-check](https://github.com/bartesto/ScatMatch/workflows/R-CMD-check/badge.svg)](https://github.com/bartesto/ScatMatch/actions)
<!-- badges: end -->

The goal of ScatMatch is to to aid error filtering from SNP genotype
data and determine group majorities through the use of visualisations.

## Installation

You can install the development version from
[GitHub](https://github.com/bartesto) with:

``` r
# install.packages("devtools")
devtools::install_github("bartesto/ScatMatch")
```

## Help Files

All functions within ScatMatch have the usual R help files, however the
best place to see these and some helpful articles describing the
intended workflow is the [ScatMatch
website](https://bartesto.github.io/ScatMatch/index.html)


NEWS.md
# ScatMatch 1.0.0

* Initial release of ScatMatch


R/hex_test.R
library(ggplot2)
library(hexSticker)




# Makes grid data and subplot for hex sticker -----------------------------

# object p must be available for plot codes to follow

# make random data for raster
d <- data.frame(expand.grid(x=1:10, y=1:10))
v <- runif(1:100)
d$v <- v

# make sublot
p <- ggplot() +
  geom_raster(data = d, aes(x=x, y=y, fill=v), alpha = 0.4) +
  scico::scale_fill_scico(palette = "vik") + #vik
  theme_void() + theme_transparent() +
  theme(legend.position = "none")

#--------------------------------------------------------------------------


# Make one plot -----------------------------------------------------------

# parameters - keep hex colours where possible
filename = XXXXXX # filename for plot including filepath (include ".png")
p_color = XXXXXX # colour for package name text
h_color = XXXXXX # colour of hex outline
h_fill = XXXXXX # colour of hex fill


hsticker <- sticker(p, package = "DBCAscatR", filename = filename,
              s_width = 1.4, s_height = 0.7, # these relate to subplot
              s_x = 1, s_y = 0.75, # these tweak location of subplot
              p_size = 22, p_color = p_color,
              h_size = 1, h_color= h_color, h_fill = h_fill)

# NOTE use plot command below to view result - if you don't it will NOT display
# as rendered
plot(hsticker)

#-------------------------------------------------------------------------


# A looped version to try a few things at once ----------------------------

# parameters
# colours to try
cols = c("#00560D", "#528D01", "#206E09", "#3BBE94", "#01A58A", "#01A58A")
# descriptors to add to filename
cnames = c("emerald", "gumdrop", "kelly", "lagoon", "mermaid", "cerulean")
# output filepath
fp = XXXXXX


for(i in seq_along(cols)){
  outfile <- paste0(fp, "/DBCAscatR_logo_", cnames[i], ".png")
  hsticker <- sticker(p, package="DBCAscatR", filename = outfile,
                s_width = 1.4, s_height = 0.7,
                s_x = 1, s_y = 0.75,
                p_size = 22, p_color = cols[i],
                h_size = 1, h_color= cols[i], h_fill = cols[i])
  plot(hsticker)
}

# can use below link to find different colours and tones
#  https://www.w3schools.com/colors/colors_picker.asp

# can add in other named colour vectors to the loop if text and borders etc are
# not the same


R/data_readiness.R
# Main data cleaning functions

#' Creates a directory structure for inputs and outputs for a `ScatMatch` workflow.
#'
#' \code{workspace} creates a directory structure for use with the `ScatMatch`
#'     functions.
#'
#' @details When run within an `RStudio` project folder structure (highly
#'     recommended), it will create  `results/` and `source/` sub-directories.
#'     After running, the USER should place a copy of the raw data input csv file to the
#'     newly created `source/` sub-directory. All outputs from `ScatMatch` functions
#'     will be written to the `results/` sub-directory.
#'
#' @return Two sub-directories, `results/` and `source/` will be created in the
#'     USERs `RStudio` project directory.
#'
#'@examples
#'\dontrun{
#'workspace()
#'}
#'
#' @author Bart Huntley, \email{bart.huntley@@dbca.wa.gov.au}
#'
#' For more details see  \url{https://bartesto.github.io/ScatMatch/index.html}
#' {the ScatMatch website}
#'
#' @import here
#'
#' @export
workspace <- function(){
  s <- here::here("source")
  r <- here::here("results")
  t_out <- here::here("results", "threshold")
  c_out <- here::here("results", "cluster")
  f_out <- here::here("results", "finalised")
  ifelse(!dir.exists(s),
         dir.create(s), FALSE)
  ifelse(!dir.exists(r),
         dir.create(r), FALSE)
  ifelse(!dir.exists(t_out),
         dir.create(t_out), FALSE)
  ifelse(!dir.exists(c_out),
         dir.create(c_out), FALSE)
  ifelse(!dir.exists(f_out),
         dir.create(f_out), FALSE)
}

#' Ingest of raw data and creation of two tidy data sets.
#'
#' \code{data_in} reads in raw data SNP genotype results file and creates two
#'     clean data sets, one with explicit NA's and the other with replacement
#'     blank cells.
#'
#' @details When run, it reads in a raw data csv file from the `source/` sub-directory
#'     performs some data cleaning and standardisation, and stores two data frames
#'     in a list. One data frame contains explicit NA's whilst the other has NA's
#'     coded as blank cells.
#'
#'     The raw data expected is SNP genotypes generated through the Agena
#'     Bioscience MassARRAY genotyping system as provided by the Australian
#'     Genome Research Facility (AGRF).
#'
#' @param filename File name of raw data input csv as a character string.
#'
#' @param replicates Logical. TRUE if replicates are present in the raw data.
#'
#' @param suffix File suffix to denote sample replicates. Default is the character
#'     string"_dup", change if required. Ignore if replicates = FALSE.
#'
#' @return When assigned to an object it will create a list containing two data
#'     frames.
#'
#'@examples
#'\dontrun{
#'data_in(filename = "CAGRF20021407_raw.csv", suffix = "_dup")
#'}
#'
#' @author Bart Huntley, \email{bart.huntley@@dbca.wa.gov.au}
#'
#' For more details see  \url{https://bartesto.github.io/ScatMatch/index.html}
#' {the ScatMatch website}
#'
#' @import here
#' @import readr
#' @importFrom janitor clean_names
#' @importFrom stringr str_detect
#' @import dplyr
#' @import tidyr
data_in <- function(filename, replicates = TRUE, suffix = "_dup"){
  suppressWarnings({
    if(replicates == TRUE){
      file <- here::here("source", filename)
      dat <- readr::read_csv(file, col_types = cols())
      # logic to catch misnamed replicate
      if(any(stringr::str_detect(dat[['X1']], suffix))){
        # data with NA
        file <- here::here("source", filename)
        na_dat <- readr::read_csv(file, na = "Fail", col_types = cols()) %>%
          janitor::clean_names() %>%
          dplyr::rename(sample = x1) %>%
          dplyr::mutate(rep = case_when(
            stringr::str_detect(sample, suffix) ~ 2,
            TRUE ~ 1),
            sample = gsub(suffix, "", sample),
            ind = stringr::str_detect(sample, "Blank")) %>%
          dplyr::filter(ind != TRUE) %>%
          dplyr::select(sample, rep, everything(), -ind)

        # data with blanks for NA
        all_dat <- read_csv(file, col_types = cols()) %>%
          janitor::clean_names() %>%
          dplyr::rename(sample = x1) %>%
          dplyr::mutate(rep = case_when(
            stringr::str_detect(sample, suffix) ~ 2,
            TRUE ~ 1),
            sample = gsub(suffix, "", sample),
            ind = stringr::str_detect(sample, "Blank")) %>%
          dplyr::filter(ind != TRUE) %>%
          dplyr::select(sample, rep, everything(), -ind) %>%
          tidyr::pivot_longer(cols = starts_with("x"),
                              names_to = "marker",
                              values_to = "val") %>%
          dplyr::mutate(val = ifelse(val == "Fail", "", val)) %>%
          tidyr::pivot_wider(names_from = marker,
                             values_from = val)

        out <- list(na_dat = na_dat, all_dat = all_dat)
        return(out)
      } else {
        stop("The suffix chosen is not present in the raw data")
      }
    } else {
      # data with NA
      file <- here::here("source", filename)
      na_dat <- readr::read_csv(file, na = "Fail", col_types = cols()) %>%
        janitor::clean_names() %>%
        dplyr::rename(sample = x1) %>%
        dplyr::mutate(rep = case_when(
          stringr::str_detect(sample, suffix) ~ 2,
          TRUE ~ 1),
          sample = gsub(suffix, "", sample),
          ind = stringr::str_detect(sample, "Blank")) %>%
        dplyr::filter(ind != TRUE) %>%
        dplyr::select(sample, rep, everything(), -ind)

      # data with blanks for NA
      all_dat <- read_csv(file, col_types = cols()) %>%
        janitor::clean_names() %>%
        dplyr::rename(sample = x1) %>%
        dplyr::mutate(rep = case_when(
          stringr::str_detect(sample, suffix) ~ 2,
          TRUE ~ 1),
          sample = gsub(suffix, "", sample),
          ind = stringr::str_detect(sample, "Blank")) %>%
        dplyr::filter(ind != TRUE) %>%
        dplyr::select(sample, rep, everything(), -ind) %>%
        tidyr::pivot_longer(cols = starts_with("x"),
                            names_to = "marker",
                            values_to = "val") %>%
        dplyr::mutate(val = ifelse(val == "Fail", "", val)) %>%
        tidyr::pivot_wider(names_from = marker,
                           values_from = val)

      out <- list(na_dat = na_dat, all_dat = all_dat)
      return(out)
    }

  })
}

#' Calculates genotyping error rates
#'
#' \code{main_errors} takes the list generated by \code{\link{data_in}} and
#'     generates genotyping error outputs which are written to csv file.
#'
#' @details When run it calculates genotyping error rates (sample amplification
#'     rate and, when replicate samples provided, allelic dropout and false allele
#'     rates) on the complete raw data set which it receives in the form of a
#'     list from the \code{\link{data_in}} function. Outputs are written to csv
#'     file to the `results/` sub-directory as:
#'     \itemize{
#'         \item numerical_alleles.csv
#'         \item sample_error_results.csv
#'         \item summary_error_results.csv}
#'
#'      These outputs are used in downstream processes and are further refined.
#'
#' @param dl data list. This is a list containing two data frames as output from
#'     the \code{\link{data_in}} function.
#'
#' @return It will write to csv file genetic error file csv's for further
#'     processing and/or evaluation.
#'
#'@examples
#'\dontrun{
#'main_errors(dl)
#'}
#'
#' @author Bart Huntley, \email{bart.huntley@@dbca.wa.gov.au}
#'
#' For more details see  \url{https://bartesto.github.io/ScatMatch/index.html}
#' {the ScatMatch website}
#'
#' @import dplyr
#' @importFrom tibble tibble
#' @import readr
#' @import tidyr
main_errors <- function(dl){
  suppressWarnings({
    na_dat <- dl[['na_dat']]
    all_dat <- dl[['all_dat']]
    # just variables
    df_loci <- na_dat  %>%
      dplyr::select(starts_with("X")) # keep loci columns for calcs

    # clean with blanks
    cl_dat_bl <- all_dat[rowSums(is.na(df_loci)) != ncol(df_loci),]

    # clean with na's
    cl_dat_na <- na_dat[rowSums(is.na(df_loci)) != ncol(df_loci),]

    # anonymous function for merge alleles for grouping step
    f1 <- function(x){trimws(paste(x, collapse = ''))}

    # set up for loop variables stuff to use in loop
    lgth_vars <- dim(df_loci)[2]
    u_samp <- unique(unlist(cl_dat_bl[, 1]))
    samp_names <- paste0("s", u_samp)

    # set up data frames to hold results
    results_out <- tibble::tibble()
    num_out <- tibble::tibble()

    # loop and calculate interim and end results
    for(i in seq_along(u_samp)){
      # using na data
      avg_amp_rate <- cl_dat_na %>%
        dplyr::filter(sample == u_samp[i]) %>%
        tidyr::pivot_longer(cols = starts_with("x"),
                            names_to = "vars",
                            values_to = "vals") %>%
        dplyr::group_by(rep) %>%
        dplyr::summarise(amp_rate = sum(!is.na(vals))/lgth_vars) %>%
        dplyr::summarise(avg_amp_rate = mean(amp_rate))

      # interim data (creating new variables to calculate errors)
      d <- cl_dat_bl %>%
        dplyr::filter(sample == u_samp[i]) %>%
        dplyr::group_by(sample) %>%
        dplyr::summarise_all(.funs = f1) %>%
        dplyr::ungroup() %>%
        tidyr::pivot_longer(cols = starts_with("x"),
                            names_to = "vars",
                            values_to = "vals") %>%
        dplyr::mutate(
          f = case_when(
            nchar(vals) == 4 ~ vals,
            TRUE ~ ""),
          c = nchar(f),
          loc_err = ifelse(c == 4 & substr(f, 1, 2) != substr(f, 3, 4),
                           0.5, 0),
          p1 = substr(f, 1, 1),
          p2 = substr(f, 2 ,2),
          p3 = substr(f, 3, 3),
          p4 = substr(f, 4 ,4),
          a_mm1 = ifelse(c == 4 & p1 != p3, 1, 0),
          a_mm2 = ifelse(c == 4 & p2 != p4, 1, 0),
          a_err = (a_mm1 + a_mm2)/4,
          het_only = case_when(
            p1 == p2 & p3 == p4 ~ 0,
            TRUE ~ 1),
          a_drop = case_when(
            het_only == 1 & substr(f, 1, 2) != substr(f, 3, 4) ~ 0.5,
            TRUE ~ 0),
          fa = case_when(
            c == 4 & het_only == 0 & substr(f, 1 ,2) != substr(f, 3 ,4) ~ 0.5,
            TRUE ~ 0))

      # take interim data and use to convert to numerical values for export
      num_vals <- d %>%
        dplyr::mutate(new_var = case_when(
          nchar(vals) == 2 ~ substr(vals, 1, 2),
          nchar(vals) == 4 & substr(vals, 1, 2) == substr(vals, 3, 4) ~ substr(vals, 1, 2),
          TRUE ~ "NA"
        )) %>%
        dplyr::mutate(to_recode = case_when(
          new_var == "NA" ~ 0,
          TRUE ~ 1
        ),
        a1 = ifelse(to_recode == 1, substr(vals, 1, 1), NA),
        b1 = ifelse(to_recode == 1, substr(vals, 2, 2), NA),
        a = case_when(
          a1 == "A" ~ 110,
          a1 == "T" ~ 120,
          a1 == "G" ~ 130,
          a1 == "C" ~ 140
        ),
        b = case_when(
          b1 == "A" ~ 110,
          b1 == "T" ~ 120,
          b1 == "G" ~ 130,
          b1 == "C" ~ 140)
        ) %>%
        dplyr::select(sample, vars, a, b) %>%
        tidyr::pivot_longer(cols = c("a", "b"),
                            names_to = "locus",
                            values_to = "value") %>%
        dplyr::arrange(vars, locus) %>%
        tidyr::pivot_wider(names_from = c(vars, locus), values_from = value)

      # take interim data and calculate errors for export
      results <- d %>%
        dplyr::summarise(locus_error = sum(loc_err)/sum(c == 4),
                         allele_error = sum(a_err)/sum(c == 4),
                         allelic_drop_out = sum(a_drop)/sum(het_only),
                         false_allele = sum(fa)/(sum(c)/4)) %>%
        dplyr::bind_cols(avg_amp_rate) %>%
        dplyr::mutate(sample = samp_names[i]) %>%
        dplyr::select(sample, avg_amp_rate, allele_error, locus_error, allelic_drop_out,
                      false_allele)

      # bind exports to appropriate result data frames
      results_out <- dplyr::bind_rows(results_out, results)
      num_out <- dplyr::bind_rows(num_out, num_vals)
    }

    # clean out errors if a no duplicate data set
    results_out_clean <- results_out %>%
      tidyr::pivot_longer(cols = -sample,
                          names_to = "error",
                          values_to = "value") %>%
      filter(!is.na(value)) %>%
      tidyr::pivot_wider(names_from = error,
                         values_from = value)

    # take the error results and further summarise
    summaries <- results_out_clean %>%
      tidyr::pivot_longer(cols = -sample,
                          names_to = "error",
                          values_to = "value") %>%
      dplyr::group_by(error) %>%
      dplyr::summarise(avg = mean(value, na.rm = TRUE),
                       se = sd(value, na.rm = TRUE)/sqrt(n()))

    # write to file errors per sample
    readr::write_csv(results_out_clean, here::here("results", "sample_error_results.csv"))

    # write to file summary errors for whole of data run
    readr::write_csv(summaries, here::here("results", "summary_error_results.csv"))

    # write to file numerical version
    readr::write_csv(num_out, here::here("results", "numerical_alleles.csv"))
  })
}

#' Main function wrapper to calculate genotyping error rates
#'
#' \code{gen_errors} takes a raw data file and generates genotyping error outputs
#'     which are written to csv file.
#'
#' @details When run it calculates genotyping errors on the complete raw data set
#'     which it reads in from the `source/` sub-directory. The raw data expected
#'     is SNP genotypes generated through the Agena Bioscience MassARRAY
#'     genotyping system as provided by the Australian Genome Research Facility
#'     (AGRF).
#'
#'     Outputs are written to csv file to the `results/` sub-directory as:
#'     \itemize{
#'         \item numerical_alleles.csv
#'         \item sample_error_results.csv
#'         \item summary_error_results.csv}
#'
#'      The genotyping error rates (sample amplification rate and, when replicate
#'      samples provided, allelic dropout and false allele rates) are used in
#'      downstream processes and are further refined.
#'
#' @inheritParams data_in
#'
#' @return It will write to csv file genotype error file csv's for further
#'     processing and/or evaluation.
#'
#'@examples
#'\dontrun{
#'gen_errors(filename = "CAGRF20021407_raw.csv", suffix = "_dup")
#'}
#'
#' @author Bart Huntley, \email{bart.huntley@@dbca.wa.gov.au}
#'
#' For more details see  \url{https://bartesto.github.io/ScatMatch/index.html}
#' {the ScatMatch website}
#'
#' @import here
#' @import readr
#' @importFrom janitor clean_names
#' @importFrom stringr str_detect
#' @importFrom tibble tibble
#' @import dplyr
#' @import tidyr
#'
#' @export
gen_errors <- function(filename, replicates = TRUE, suffix = "_dup"){
  suppressWarnings({
    # ingest data and make data list
    dl <- data_in(filename, replicates, suffix)
    # munge and create error list
    main_errors(dl)
  })
}

#' Applies quality filtering of samples by choosing an amplification threshold
#' below which samples are discarded and produces interim filtered results.
#'
#' \code{amp_threshold} takes a threshold value and applies it to the numerical
#'     allele data. Also produces a record of which samples were filtered out and
#'     proportions of NA's per loci.
#'
#' @details The predetermined amplification threshold, ascertained through
#'     visualisation \code{\link{amp_splots}}, is used to filter the numerical
#'     alleles data set and written to csv. The names of all the samples that are
#'     rejected at this threshold are also written to csv file.
#'
#'     Lastly the proportion of missing data (NA’s) per locus is written to csv.
#'     All outputs are written to `results/`.
#'
#' @param at Numeric value of the average amplification rate to filter the
#'     numerical alleles. Values greater than or equal to are retained.
#'
#' @return It will write to `results/` three csv files:
#'     \itemize{
#'         \item filtered alleles data with threshold indicated in the name
#'         \item sample that were filtered at with threshold indicated in the name
#'         \item proportion of NA's for the loci}
#'
#'     It will also print to screen the number of samples filtered out at this
#'     threshold.
#'
#'@examples
#'\dontrun{
#'amp_threshold(at = 0.8)
#'}
#'
#' @author Bart Huntley, \email{bart.huntley@@dbca.wa.gov.au}
#'
#' For more details see  \url{https://bartesto.github.io/ScatMatch/index.html}
#' {the ScatMatch website}
#'
#' @import here
#' @import readr
#' @import dplyr
#' @import tidyr
#' @importFrom stringr str_replace
#'
#' @export
amp_threshold <- function(at){
  suppressWarnings({
    results_out <- readr::read_csv(here::here("results",
                                              "sample_error_results.csv"),
                                   col_types = cols())
    num_out <- readr::read_csv(here::here("results",
                                          "numerical_alleles.csv"),
                               col_types = cols())
    # filter results list by amplification success rate
    filt_out <- results_out %>%
      dplyr::filter(avg_amp_rate >= at)

    filt_names <- filt_out['sample']
    filt_names[['sample']] <- stringr::str_replace(filt_names[['sample']], "s",
                                                   "")
    # filter num_out dataframe by sample name values in filt_out
    num_out_filt_samps <- num_out[num_out[['sample']] %in% filt_names[['sample']], ]

    # write to file interim filtered alleles on average amplification rate (at)
    cname <- paste0("numerical_alleles_filtered_at", at, ".csv")
    readr::write_csv(num_out_filt_samps, here::here("results", "threshold", cname))

    # for records, get a list of samples that were filtered
    samples_x <- results_out %>%
      dplyr::filter(avg_amp_rate < at) %>%
      dplyr::select(sample, avg_amp_rate)

    #print to screen number of removals
    ls <- length(samples_x[['sample']])
    cat("At an amplification threshold  of", at, ": filtering out", ls, "samples")

    # write to file samples filtered out based on average amplification rate (at)
    c2name <- paste0("samples_filtered_at", at, ".csv")
    readr::write_csv(samples_x, here::here("results", "threshold", c2name))

    # summarise proportion NAs per locus in filtered num_out dataframe
    loc_NA <- num_out_filt_samps %>% summarise_all(~ mean(is.na(.))) %>%
      dplyr::select(-sample) %>%
      tidyr::pivot_longer(everything(), names_to = "loci",
                          values_to = "proportion")

    # write to file props of NA per locus for histogram plotting
    readr::write_csv(loc_NA, here::here("results", "threshold", "loci_NA.csv"))

  })
}

#' Applies quality filtering of loci by applying a maximum amount of missing data
#' allowed per locus above which loci are discarded
#'
#' \code{miss_threshold} takes a "missingness" threshold and applies it to a
#'     filtered numerical allele data set and then writes the results to csv.
#'
#' @details The predetermined "missingness" threshold, ascertained through
#'     visualisation \code{\link{miss_hist}}, is used to further filter the
#'     numerical alleles data set and written to csv to the `results/`
#'     sub-directory.
#'
#'     The amplification threshold parameter is used to select the correct
#'     data set from the `results/` sub-directory that had been created by the
#'     \code{\link{amp_threshold}} function.
#'
#' @param mt Numeric value of the "missingness" threshold to further filter the
#'     numerical alleles. Values less than the threshold are retained.
#'
#' @param at Numeric value of the average amplification threshold used on the
#'     input data. NOTE used to select a pre-made data.
#'
#' @return Writes a numerical alleles data csv, filtered to the `at` and `mt`
#'     thresholds, to the `results/` sub-directory. It will also print to screen
#'     the number of loci that will be filtered out at this "missingness" threshold.
#'
#'@examples
#'\dontrun{
#'miss_threshold(mt = 0.2, at = 0.8)
#'}
#'
#' @author Bart Huntley, \email{bart.huntley@@dbca.wa.gov.au}
#'
#' For more details see  \url{https://bartesto.github.io/ScatMatch/index.html}
#' {the ScatMatch website}
#'
#' @import here
#' @import readr
#' @import dplyr
#'
#' @export
miss_threshold <- function(mt, at){
  suppressWarnings({

    # read in locus NA proportion data
    loc_NA_names <- readr::read_csv(here::here("results", "threshold", "loci_NA.csv"),
                                    col_types = cols()) %>%
      dplyr::filter(proportion >= mt) %>%
      dplyr::pull(loci)

    #print to screen number of removals
    ln <- length(loc_NA_names)
    cat("At an missingness threshold of", mt, ": filtering out", ln, "loci")

    # make name of correct amplification filtered data and read in
    num_name <- paste0("numerical_alleles_filtered_at", at, ".csv")
    num_out_filt_samps <-  readr::read_csv(here::here("results", "threshold",
                                                      num_name),
                                           col_types = cols())

    # filter loci with high missingness from num_out dataframe
    num_out_filt_loci <- num_out_filt_samps %>%
      dplyr::select(-all_of(loc_NA_names))

    # write to file numerical version that has been quality filtered for samples
    # and loci
    cname <- paste0("numerical_alleles_filtered_at", at, "_mt", mt, ".csv")
    readr::write_csv(num_out_filt_loci, here::here("results", "threshold", cname))

  })
}


R/hex.R
library(ggplot2)
library(hexSticker)
library(scico)
library(here)

# make random data for raster
d <- data.frame(expand.grid(x=1:10, y=1:10))
v <- runif(1:100)
d$v <- v

# make sublot
p2 <- ggplot() +
  geom_raster(data = d, aes(x=x, y=y, fill=v), alpha = 0.4) +
  scico::scale_fill_scico(palette = "vik") + #vik
  theme_void() + theme_transparent() +
  theme(legend.position = "none")


# output name
cols = c("#00560D", "#528D01", "#206E09", "#3BBE94", "#01A58A", "#01A58A")
cnames = c("emerald", "gumdrop", "kelly", "lagoon", "mermaid", "cerulean")




for(i in seq_along(cols)){
  outfile <- here("man", "figures", paste0("DBCAscatR_logo_", cnames[i], ".png"))
  t2 <- sticker(p2, package="DBCAscatR", filename = outfile,
                s_width = 1.4, s_height = 0.7,
                s_x = 1, s_y = 0.75,
                p_size = 22, p_color = "darkred",
                h_size = 1, h_color= "darkred", h_fill = cols[i])#"#003399" "snow2"
  plot(t2)
}

# find different tones at https://www.w3schools.com/colors/colors_picker.asp
# cerulean "#01A58A"
cshades = c("#01987f", "#017f6a", "#016555", "#004c3f")
cgrade = c("30", "25", "20", "15")

for(i in seq_along(cshades)){
  outfile <- here("man", "figures", paste0("DBCAscatR_logo_", cgrade[i], ".png"))
  t2 <- sticker(p2, package="DBCAscatR", filename = outfile,
                s_width = 1.4, s_height = 0.7,
                s_x = 1, s_y = 0.75,
                p_size = 22, p_color = cshades[i],
                h_size = 1, h_color= cshades[i], h_fill = "#01A58A")#"#003399" "snow2"
  plot(t2)
}





t2 <- sticker(p2, package="DBCAscatR", filename=outfile,
              s_width = 1.4, s_height = 0.7,
              s_x = 1, s_y = 0.75,
              p_size = 22, p_color = "lagoon",
              h_size = 1, h_color= "darkred", h_fill = "rosybrown3")#"#003399" "snow2"

# use this to view as viewer and png very different - plot call will view as
# png
plot(t2)
